core/src/types/list.rs

```diff
@@ -80,43 +83,87 @@ impl Stream for Lister {
             return Poll::Ready(None);
         }
 
-        if let Some(fut) = self.stating.as_mut() {
-            let (path, rp) = ready!(fut.poll_unpin(cx));
-
-            // Make sure we will not poll this future again.
-            self.stating = None;
-            let metadata = match rp {
-                Ok(rp) => rp.into_metadata(),
-                Err(err) => {
-                    self.errored = true;
-                    return Poll::Ready(Some(Err(err)));
-                }
-            };
-
-            return Poll::Ready(Some(Ok(Entry::new(path, metadata))));
+        if !self.task_queue.is_empty() {
+            if let Some(handle) = self.task_queue.back_mut() {
+                let (path, rp) = ready!(handle.poll_unpin(cx)).map_err(|err| {
+                    Error::new(ErrorKind::Unexpected, "join handle error")
+                        .with_operation("types::Lister::poll_next")
+                        .set_source(err)
+                })?;
+
+                return match rp {
+                    Ok(rp) => {
+                        self.task_queue.pop_back();
+                        let metadata = rp.into_metadata();
+                        Poll::Ready(Some(Ok(Entry::new(path, metadata))))
+                    }
+                    Err(err) => {
+                        self.errored = true;
+                        Poll::Ready(Some(Err(err)))
+                    }
+                };
+            }
         }
 
-        match ready!(self.lister.poll_next(cx)) {
-            Ok(Some(oe)) => {
-                let (path, metadata) = oe.into_entry().into_parts();
-                if metadata.contains_metakey(self.required_metakey) {
-                    return Poll::Ready(Some(Ok(Entry::new(path, metadata))));
-                }
-
+        if let Some(entry) = self.buf.clone() {
+            let (path, metadata) = entry.into_parts();
+            return if metadata.contains_metakey(self.required_metakey) {
+                self.buf = None;
+                Poll::Ready(Some(Ok(Entry::new(path, metadata))))
+            } else {
                 let acc = self.acc.clone();
                 let fut = async move {
                     let res = acc.stat(&path, OpStat::default()).await;
-
                     (path, res)
                 };
-                self.stating = Some(Box::pin(fut));
+
+                self.task_queue.push_front(tokio::spawn(fut));
+                self.buf = None;
                 self.poll_next(cx)
-            }
-            Ok(None) => Poll::Ready(None),
-            Err(err) => {
-                self.errored = true;
-                Poll::Ready(Some(Err(err)))
-            }
+            };
+        }
+
+        loop {
```

Comment:
Current implemention doesn't have concurrent. In the worst case, we will just have only one `JoinHandle`:

- `lister.poll_next` returns `Polling::Ready(Ok(Some(entry))`, pushed into `task_queue`.
- The next `lister.poll_next` returns `Polling::Pending`, so we exit until next user call.
- task queue is not empty now, we are polling the `task_queue.back_mut`
- After the join handle in `task_queue` returns `Ready`, we start over.

In this way, no concurrent will happen since we only have on join handle created.

The root cause is that we should handle `ready!()` carefully.

---

Maybe we can call `self.lister.poll_next(cx)` first to fill the `taks_queue`, than we can poll the join handle in task queue?

Reply:
Ah... I saw it ! 
https://github.com/apache/incubator-opendal/blob/925c8e82e2e8e4c4ed3260f3c7994c338c598947/core/src/raw/oio/list/page_list.rs#L127

Moved `self.lister.poll_next(cx)` to the top.
By the way, I think I need some guidance regarding the `buf` problem. If the `metakey` is already known and the `task_queue` is not empty, what should we do then?

Reply:
> By the way, I think I need some guidance regarding the `buf` problem. If the `metakey` is already known and the `task_queue` is not empty, what should we do then?

Nice catch. 

I believe we should store an enum in `task_queue`, for example:

```rust
enum Abc {
  X(JoinHandle),
  Y(Option<(String, RpStat)>)
}
```

(nameing is on your decision)

In this way, we can return the already known entry directly without an extra poll.

Reply:
Just a note, all queue-related code in opendal uses `push_back` and `pop_front`. Would you prefer to adjust the code in this PR for consistency?

Reply:
changed to `push_back` and `pop_front`.